File: ruby/practice/classic_puzzle/easy/minicpu-instruction-decoder.rb
# Decode and execute the MiniCPU bytecode program.

program = gets.chomp # Space-separated hex bytes representing CPU instructions

# Write an answer using puts
# To debug: STDERR.puts "Debug messages..."

class MiniCpu
  REGISTER_NUMBERS = %w[00 01 02 03].freeze

  def self.exec(program)
    @register = REGISTER_NUMBERS.to_h { |key| [key, 0] }

    bytes = program.split
    while instruction = Instruction.from_code(bytes.shift)
      break if instruction == Instruction::HLT

      if [Instruction::INC, Instruction::DEC].include? instruction
        register_number = bytes.shift
        method(instruction.operation).call register_number
      elsif [Instruction::MOV, Instruction::ADD, Instruction::SUB, Instruction::MUL].include? instruction
        register_number, value = bytes.shift 2
        value = value.to_i 16 if instruction == Instruction::MOV

        method(instruction.operation).call register_number, value
      end
    end

    @register.transform_values { |value| value % 256 }.values_at(*REGISTER_NUMBERS)
  end

  def self.mov(register_number, value)
    @register[register_number] = value
  end

  def self.add(register_number1, register_number2)
    @register[register_number1] += @register[register_number2]
  end

  def self.sub(register_number1, register_number2)
    @register[register_number1] -= @register[register_number2]
  end

  def self.mul(register_number1, register_number2)
    @register[register_number1] *= @register[register_number2]
  end

  def self.inc(register_number)
    @register[register_number] += 1
  end

  def self.dec(register_number)
    @register[register_number] -= 1
  end

  class Instruction
    attr_reader :code, :operation

    def initialize(code, operation)
      @code = code
      @operation = operation
    end

    MOV = new '01', :mov
    ADD = new '02', :add
    SUB = new '03', :sub
    MUL = new '04', :mul
    INC = new '05', :inc
    DEC = new '06', :dec
    HLT = new 'FF', nil

    def self.from_code(code)
      [MOV, ADD, SUB, MUL, INC, DEC, HLT].find { |instruction| instruction.code == code }
    end
  end
end

results = MiniCpu.exec program

# Print the final value of each register R0, R1, R2, R3, one value per line
# puts "answer"
results.each do |result|
  puts result
end

File: typescript/practice/classic_puzzle/easy/minicpu-instruction-decoder.ts
/**
 * Decode and execute the MiniCPU bytecode program.
 **/

const program: string = readline(); // Space-separated hex bytes representing CPU instructions

// Write an answer using console.log()
// To debug: console.error('Debug messages...');

class MiniCpu {
    static BIT_SIZE: number = 256;

    static register: { [key: string]: number };

    static exec: (program: string) => string[] = (program: string): string[] => {
        type Instruction = {
            code: string,
            operation: (...args: [registerNumber: string] | [registerNumber: string, value: number] | [registerNumber1: string, registerNUmber2: string]) => void,
        }
        const instructions: { [key: string]: Instruction } = {
            mov: {
                code: '01',
                operation: this.mov,
            },
            add: {
                code: '02',
                operation: this.add,
            },
            sub: {
                code: '03',
                operation: this.sub,
            },
            mul: {
                code: '04',
                operation: this.mul,
            },
            inc: {
                code: '05',
                operation: this.inc,
            },
            dec: {
                code: '06',
                operation: this.dec,
            },
            hlt: {
                code: 'FF',
                operation: null,
            },
        };

        const registerNumbers: string[] = ['00', '01', '02', '03'];

        this.register = {};
        registerNumbers.forEach(key => this.register[key] = 0);

        const bytes: string[] = program.split(' ');
        while (bytes.length > 0) {
            const code: string = bytes.shift();
            const instruction: Instruction = Object.values(instructions).find(instruction => instruction.code === code);
            if (instruction === instructions.hlt) {
                break;
            }

            if (instruction === instructions.mov) {
                const registerNumber: string = bytes.shift();
                const valueStr: string = bytes.shift();
                const value: number = parseInt(valueStr, 16);

                instruction.operation(registerNumber, value);
            } else if ([instructions.add, instructions.sub, instructions.mul].includes(instruction)) {
                const registerNumber: string = bytes.shift();
                const value: string = bytes.shift();

                instruction.operation(registerNumber, value);
            } else if ([instructions.inc, instructions.dec].includes(instruction)) {
                const registerNumber: string = bytes.shift();
                instruction.operation(registerNumber);
            }
        }

        return registerNumbers.map(registerNumber => {
            return (this.register[registerNumber] % 256).toString();
        });
    };

    static mov: (registerNumber: string, value: number) => void = (registerNumber: string, value: number): void => {
        this.register[registerNumber] = value;
    };

    static add: (registerNumber1: string, registerNumber2: string) => void = (registerNumber1: string, registerNumber2: string): void => {
        this.register[registerNumber1] = (this.register[registerNumber1] + this.register[registerNumber2]) % this.BIT_SIZE;
    };

    static sub: (registerNumber1: string, registerNumber2: string) => void = (registerNumber1: string, registerNumber2: string): void => {
        this.register[registerNumber1] = (this.register[registerNumber1] - this.register[registerNumber2] + this.BIT_SIZE) % this.BIT_SIZE;
    };

    static mul: (registerNumber1: string, registerNumber2: string) => void = (registerNumber1: string, registerNumber2: string): void => {
        this.register[registerNumber1] = (this.register[registerNumber1] * this.register[registerNumber2]) % this.BIT_SIZE;
    };

    static inc: (registerNumber: string) => void = (registerNumber1: string): void => {
        this.register[registerNumber1] = (this.register[registerNumber1] + 1) % this.BIT_SIZE;
    };

    static dec: (registerNumber: string) => void = (registerNumber1: string): void => {
        this.register[registerNumber1] = (this.register[registerNumber1] - 1 + this.BIT_SIZE) % this.BIT_SIZE;
    };
}

const results: string[] = MiniCpu.exec(program);

// Print the final value of each register R0, R1, R2, R3, one value per line
// console.log('answer');
results.forEach(result => {
    console.log(result);
});
